### ISSUE TITLE
ruby cucumber: --require flag documentation appears to be incorrect

### ISSUE THREAD

**Issue by fourpastmidnight** [cucumber/cucumber#567]
## Summary

I'm working on improving the PowerShell Pester implementation of their Gherkin test runner such that it is more like cucumber as used with Ruby. I've closely read the `cucumber --help` output and began running some simple tests to see how cucumber discovered feature files, `env.rb` files, and support Ruby scripts. (I've also read _The Cucumber Book_ to also help me--though the book is a bit dated now, but still a great reference! Thanks @aslakhellesoy and @mattwynne!)

The output of `cucumber --help` as it relates tot he `--require` flag is:

> Require files before executing the features. If this option is not specified, all *.rb files that are siblings or below the features will be loaded automatically. Automatic loading is disabled when this option is specified, and all loading becomes explicit. _Files under directories named "support" are always loaded first._

(Emphasis mine) I take this to mean that even if you use `--require`, any `*.rb` files under the `support` directory are loaded.

Also, realizing that _The Cucumber Book_ is a bit dated, on page 130 it says:

> The file `features/support/env.rb` is always the very first file to be loaded when Cucumber starts a test run.

I have the following on-disk structure:

```text
calculator
  |--- features
       |--- adding.feature
       |--- foo
       |     |--- support
       |     |    \--- env.rb    # puts "Loaded foo env.rb"
       |     \--- utility.rb     # puts "Loaded foo utility.rb"
       |--- step_definitions
       \--- support
                \--- env.rb      # puts "Loaded env.rb"
```

Here are some commands I ran, as well as some (abbreviated--not showing output from loaded feature files) output:

```shell
$ cucumber --verbose
Code:
  * features/foo/support/env.rb
Loaded feature foo env.rb
  * features/support/env.rb
Loaded env.rb

Features:
  * features/adding.feature
```

So, the above output seems to contradict _The Cucumber Book_, as it would appear that `features/foo/support/env.rb` was loaded _before_ `features/support/env.rb`.

```shell
$ cucumber --require features/foo/utility.rb --verbose
Code:
  * features/foo/utility.rb
Loaded foo utility.rb

Features:
  * features/adding.feature
```

And here in the above example, the use of the `--require` flag seems to contradict the content of `--help` because `features/support/env.rb` was not loaded at all!

## Expected Behavior

I expected that `features/support/env.rb` would be loaded first in both examples above.

## Current Behavior

If other Ruby scripts exist in other support directories, they seem to be loaded before `features/support/env.rb` in contradiction to _The Cucumber Book_, p. 130. If `--require` is used, then `features/support/env.rb` isn't loaded at all, in contradiction to the text displayed by `--help` for that option.

## Your Environment

* Version used: Ruby Cucumber 3.1.2
* Operating System and version: Windows 10, October 2018 Update (Ubuntu WSL)
* Link to your project: N/A

**Comment by danascheider** [comment]
@fourpastmidnight You are correct - when the `--require` flag is used, only the directories passed to that flag are loaded. If you need another file loaded, you'll need to `--require` it separately. Note that the `--help` output says
> Files under _directories named_ support are always loaded first.
The meaning of this is that files under directories named `support` _in the directory tree `--require`d_ will be loaded first.

**Comment by fourpastmidnight** [comment]
@danascheider AH! Thanks. OK. So, that is another way I considered "interpreting" that sentence, but opted for my first interpretation. So, I guess the ask here is that the help for that option be cleaned up a bit to make it less ambiguous.

And, I suppose, then, that the quote from _The Cucumber Book_ is either a) outdated or b) unintentionally mis-leading or incorrect? That `--require`, does as it's help text says, it makes _all_ automatic loading explicit, even for `./feature/support/env.rb`?

**Comment by fourpastmidnight** [comment]
Sorry, I had too many questions in this issue 😞  But one final clarification, too: the order of loading of support scripts. _The Cucumber Book_ also made it seem as if when `./features/support/env.rb` is loaded, it is _always_ loaded _first_. But my output does not seem to indicate that is true. It seemed to me that the impetous behind loading `./features/support/env.rb` first was to guarantee that whatever is in that file is imported into the environment since other scripts may depend on things in that file. Otherwise, if loading is random--or as appears to be in this case a depth-first traversal of the filesystem under `./features`--you'd have quite a dependency mess on your hands (of course, if you have that many dependencies between support files, then you may have other problems, too. 😉).

**Comment by fourpastmidnight** [comment]
@danascheider I answered my own question about `env.rb` with respect to `--require`. It seems I read too much into the word _always_. That word only applies when `--require` is not used and automatic loading is in effect and cucumber is loading files found under `./features/support`.

However, I'm still curious as to the asnwer to my other question immediately above with respect to load order.

**Comment by mattwynne** [comment]
Hi @fourpastmidnight thanks for taking an interest in this. The code you're trying to decipher is even older than that dated Cucumber Book you've got, and you're not the first person to find it's behaviour confusing. I'd love any concrete suggestions for how we could clean it up.

To clarify that quote from the book (and as Dana has also said this above) as soon as you use `--require` all the automatic loading is disabled, and we just load stuff from the path(s) you've passed us on the command line. If you don't use `--require` then it does this stuff with trying to pick a `support/env.rb` first and so on.

**Comment by fourpastmidnight** [comment]
Yep, I realized with respect to `--require` that I took the quote out of context. Thanks!

What about the load order of “support” files, in general, _without_ the use of `--require`? The quote says, in as  many words, that the `./features/support/env.rb` file is _always_ loaded first. But in my example above, where `--require` *is not* used, that file appears to actually be loaded last!

This is the only remaining behavior that seems counter to what I’m reading that I wish to ensure I understand correctly.

Sent from my Windows 10 phone

**Comment by mattwynne** [comment]
Honestly Craig, I don't know. I'm surprised by what you've observed since it's counter to my current mental model, but not entirely. It's awkward to test this kind of code and I don't think it does have brilliant tests at the moment, so it could well be doing something unexpected if you have another `support/env.rb` file somewhere else.

Do you have the patience / time / enthusiasm to help us to pin this down with some scenarios, and potentially change the behaviour to be more like what you'd expect?

**Comment by fourpastmidnight** [comment]
Thanks Matt. I'll do some poking around in the cucumber code and see what I can find. I haven't done Ruby code in over 20 years (back when it was 0.4 or something like that!) But, I'm interested to see what the real behavior of the code is anyway, especially if I'm more or less "porting" it to PowerShell. If I find something, I'll report back.

**Comment by fourpastmidnight** [comment]
OK, I cloned the `cucumber-ruby` repository because I was looking at the tests to help guide my PowerShell implementation and I found the test that looked to be the one I would need to modify to see if there truly is a problem.

```ruby
# in spec/cucumber/configuration_spec.rb, line 32:
it 'requires env.rb files first' do
  configuration = Configuration.new
  # Below is the original line. And below that is the new line which makes the test fail.
  # I included two levels of support folders to prove my hypothesis that the tree node
  # visit strategy is LVR when it should be VLR.
  #given_the_following_files('/features/support/a_file.rb', '/features/support/env.rb')
  given_the_following_files(
    '/features/foo/bar/support/a_file.rb',
    '/features/foo/bar/support/env.rb',
    '/features/foo/support/a_file.rb',
    '/features/foo/support/env.rb',
    '/features/support/a_file.rb',
    '/features/support/env.rb')

  # And below is the original expectation and below that the new one
  #expect(configuration.support_to_load).to eq [
  #  '/features/support/env.rb',
  #  '/features/support/a_file.rb'
  #]
  expect(configuration.support_to_load).to eq [
    '/features/support/env.rb',
    '/features/foo/support/env.rb',
    '/features/foo/support/bar/env.rb',
    '/features/support/a_file.rb',
    '/features/foo/support/a_file.rb',
    '/features/foo/bar/support/a_file.rb'
  ]
end
```

Once I changed the test above as shown, the test failed. Here's the output (note, I line wrapped the expected/got output so it's easier to read in this issue):

```text
Failures:

  1) Cucumber::Configuration selecting files to load requires env.rb files first
     Failure/Error:
       expect(configuration.support_to_load).to eq [
         '/features/support/env.rb',
         '/features/foo/support/env.rb',
         '/features/foo/bar/support/env.rb',
         '/features/support/a_file.rb',
         '/features/foo/support/a_file.rb',
         '/features/foo/bar/support/a_file.rb'
       ]

       expected: [
                "/features/support/env.rb",
                "/features/foo/support/env.rb",
                "/features/foo/bar/support/env.rb",
                "/features/support/a_file.rb",
                "/features/foo/support/a_file.rb",
                "/features/foo/bar/support/a_file.rb"
           ]
            got: [
                "/features/foo/bar/support/env.rb",
                "/features/foo/support/env.rb",
                "/features/support/env.rb",
                "/features/foo/bar/support/a_file.rb",
                "/features/foo/support/a_file.rb",
                "/features/support/a_file.rb"
            ]

       (compared using ==)
     # ./spec/cucumber/configuration_spec.rb:42:in `block (3 levels) in <module:Cucumber>'

Finished in 13.37 seconds (files took 1.78 seconds to load)
470 examples, 1 failure, 1 pending

Failed examples:

rspec ./spec/cucumber/configuration_spec.rb:32 # Cucumber::Configuration selecting files to load requires env.rb files first

/usr/bin/ruby2.5 -I/var/lib/gems/2.5.0/gems/rspec-core-3.8.0/lib:/var/lib/gems/2.5.0/gems/rspec-support-3.8.0/lib /var/lib/gems/2.5.0/gems/rspec-core-3.8.0/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb failed
```

This is good! This is what I expected, that Cucumber is performing a LVR tree node visit strategy when walking the filesystem looking for `env.rb` files. I think if we change the visit stratgey to VLR, then the files would be found and loaded in the correct order, as indicated by _The Cucumber Book_ on page 129ff (though, _The Cucumber Book_ does not explicitly state what the behavior should be for the set of files as specified in the modified test above):

> Just like `features/step_definitions`, Cucumber will load all Ruby files it finds in `features/support`. This is convenient, because it means you don't need to pepper `require` statements everywhere, but it does mean that you can't have dependencies between files in the `support` directory, because you can't expect that another file will have already been loaded. There is one exception to that, a special file called `env.rb`.
>
> The file `features/support/env.rb` is always the very first file to be loaded when Cucumber starts a test run. You use it to prepare the environment for the rest of your support and step definition code to operate.

However, from the above, it seems it really doesn't matter in what order other support files which are found are loaded, as files in the various `support` directories should not have any dependencies due to the undeterministic load ordering. However, it would seem that standardizing on a VLR visit strategy would make the most sense because both visit strategies can be used for loading all support files--just the `env.rb` files would be searched for and loaded first.

Obviously, the above assumes that `--require` has not been specified, pursuant to our discussion earlier in this issue.

**Comment by fourpastmidnight** [comment]
Turns out the solution was _really_ simple: just reverse the order of the files in `env_files` and `other_files` it `configuration.rb` on line 177. I'll submit a PR.

**Comment by fourpastmidnight** [comment]
[Here's the pull request.](https://github.com/cucumber/cucumber-ruby/pull/1343?ts=2)

**Comment by fourpastmidnight** [comment]
And [here's the pull request](https://github.com/cucumber/cucumber-ruby/pull/1344) to update the documentation for the `--require` option. I've already gotten good feedback on this.

**Comment by olleolleolle** [comment]
@fourpastmidnight I have merged the `env.rb` file sorting PR, too!

**Comment by fourpastmidnight** [comment]
Thanks! I really wanted some eyes on it in case anyone felt strongly that the change in behavior would hurt more than help. But at least Cucumber now behaves as documented.

Thanks again!

Sent from my Windows 10 phone

**Comment by luke-hill** [comment]
I'm going to close this as fixed @fourpastmidnight.

Currently the master branch is tracking a v4.rc1 (Of sorts), but this will be released in the next cut of v4 (And a v4.0 of course)

Thanks for your help

**Comment by fourpastmidnight** [comment]
Thanks. I forgot to do that.

Sent from my Windows 10 phone
